specs/lazy-adr/adr-021-lazy-aggregation.md

```diff
@@ -0,0 +1,160 @@
+
+## Alternative Approaches
+
+### 1. Skip Empty Blocks Entirely
+
+- Continue current behavior of not posting empty blocks
+- **Rejected** because it breaks EVM execution layer expectations and creates inconsistent block mapping
+
+### 2. Post Full Block Structure for Empty Blocks
+
+- Post regular blocks with empty transaction lists
+- **Rejected** because it wastes DA layer space unnecessarily
+
+### 3. Sentinel Empty Block Marker (Chosen)
+
+- Post a special single-byte transaction to mark empty blocks
+- Optimize DA layer space usage while maintaining block consistency
+- Provides clear distinction between empty and failed/missed blocks
+
+## Decision
+
+Implement a sentinel-based empty block marking system that:
+
+1. Posts a special sentinel transaction to the DA layer for empty blocks
+2. Maintains block height consistency between DA and execution layers
+3. Optimizes DA layer space usage
+4. Preserves existing lazy aggregation timing behavior
+
+## Detailed Design
+
+### User Requirements
+
+- Maintain consistent block heights between DA and execution layers
+- Minimize DA layer costs for empty blocks
+- Preserve existing lazy aggregation performance characteristics
+- Support EVM execution layer expectations
+
+### Systems Affected
+
+1. Block Manager (`block/aggregation.go`)
+2. Transaction Processing Pipeline
+3. DA Layer Interface
+4. Block Execution Logic
+
+### Data Structures
+
+```go
+const (
+    // EmptyBlockSentinel is a special byte sequence marking empty blocks
+    EmptyBlockSentinel = []byte{0x00, 0xE0, 0x0E} // "EMPTY" in minimal form
+)
+
+type BlockData struct {
+    Txs [][]byte
+    IsEmpty bool
+}
+```
+
+### Implementation Details
+
+1. **Empty Block Detection**:
+
+    ```go
+    func (m *Manager) isEmptyBlock(txs [][]byte) bool {
+        return len(txs) == 1 && bytes.Equal(txs[0], EmptyBlockSentinel)
+    }
+    ```
+
+2. **Modified Aggregation Loop**:
+
+    ```go
+    func (m *Manager) lazyAggregationLoop(ctx context.Context, blockTimer *timTimer) {
+        // ... existing timer setup ...
+
+        for {
+            select {
+            case <-ctx.Done():
+                return
+            case <-lazyTimer.C:
+                // No transactions available, create empty block
+                if err := m.publishEmptyBlock(ctx); err != nil {
+                    m.logger.Error("error publishing empty block", "error", err)
+                }
+            case <-blockTimer.C:
+                // Normal block production with available transactions
+                if err := m.publishBlock(ctx); err != nil {
+                    m.logger.Error("error publishing block", "error", err)
+                }
+            }
+        }
+    }
+
+    func (m *Manager) publishEmptyBlock(ctx context.Context) error {
+        return m.publishBlock(ctx, EmptyBlockSentinel)
```

Comment:
minor nit: golang doesn't support method overloading so we'd need a separate function since there's already a `publishBlock` that takes one argument `context`😅